### PR TITLE
decomp: almost all of the NPCs!

### PR DESCRIPTION
**Pull request by xTVaser**
- mayor
- bird-lady
- bird-lady-beach
- sculptor
- geologist
- oracle
- farmer
- explorer
- assistant (blocked by sp-launch-group from static data)
- sage
- gambler
- warrior
- miners (blocked by sp-launch-group)

### CODE REVIEW

File: goal_src/levels/beach/mayor.gc [hunk 1]
```diff
@@ -5,3 +5,2238 @@
...
+   (< (- (-> (camera-pos) z) (-> obj root-override trans z)) 57344.0)
+   )
+  )
+
+;; definition for method 43 of type mayor
+(defmethod TODO-RENAME-43 mayor ((obj mayor))
+  (when
+   (TODO-RENAME-10
+    (-> obj ambient)
+    (the-as vector (new 'stack-no-clear 'ambient-control))
```
Comment (water111) [inline goal_src/levels/beach/mayor.gc:2056]: I think the stack structure should be a `vector` here

File: goal_src/levels/beach/mayor.gc [hunk 2]
```diff
@@ -5,3 +5,2238 @@
...
+         (-> obj tasks)
+         (game-task jungle-lurkerm)
+         (task-status need-reminder)
+         )
+        )
+       (-> obj ambient)
+       (let ((t9-3 (method-of-type ambient-control dummy-11)))
+        "CHI-LO01"
+        (-> obj root-override trans)
+        (t9-3)
```
Comment (water111) [inline goal_src/levels/beach/mayor.gc:2079]: method args again (I think it's the same one as before)

File: goal_src/levels/beach/sculptor.gc [hunk 3]
```diff
@@ -5,3 +5,1314 @@
...
+          (v1-4 (the-as number (logior #x3f800000 v1-3)))
+          (f0-2 (+ -1.0 (the-as float v1-4)))
+          )
+    (cond
+     ((< 0.85714287 f0-2)
+      (-> obj ambient)
+      (let ((t9-2 (method-of-type ambient-control dummy-11)))
+       "SCU-LO01"
+       (-> obj root trans)
+       (t9-2)
```
Comment (water111) [inline goal_src/levels/beach/sculptor.gc:418]: method again

File: goal_src/levels/beach/sculptor.gc [hunk 4]
```diff
@@ -5,3 +5,1314 @@
...
+       )
+      (set! (-> a0-12 param 1) 1.0)
+      (set! (-> a0-12 frame-num) 0.0)
+      (joint-control-channel-group!
+       a0-12
+       (the-as art-joint-anim (-> self draw art-group data 12))
+       num-func-seek!
+       )
+      )
+     (until (begin
```
Comment (water111) [inline goal_src/levels/beach/sculptor.gc:539]: hopefully if you run again with the changes in https://github.com/water111/jak-project/pull/768, this loop will look a lot nicer

File: goal_src/levels/beach/sculptor.gc [hunk 5]
```diff
@@ -5,3 +5,1314 @@
...
+         )
+        (set! (-> a0-18 param 1) 1.0)
+        (set! (-> a0-18 frame-num) 0.0)
+        (joint-control-channel-group!
+         a0-18
+         (the-as art-joint-anim (-> self draw art-group data 8))
+         num-func-seek!
+         )
+        )
+       (until (begin
```
Comment (water111) [inline goal_src/levels/beach/sculptor.gc:592]: same with this one

File: goal_src/levels/village1/sage.gc [hunk 6]
```diff
@@ -5,3 +5,1422 @@
...
+   ((-> (method-of-type process-taskable play-anim) exit))
+   (case (get-task-status (game-task intro)) 
+    (((task-status need-resolution))
+      (clear-pending-settings-from-process
+       *setting-control*
+       self
+       'music-volume-movie
+       )
+      (close-specific-task! (game-task intro) (task-status need-resolution))
+      (start 'play (get-continue-by-name *game-info* "game-start"))
```
Comment (water111) [inline goal_src/levels/village1/sage.gc:1354]: Cool - I think this is the function call that starts the game.

File: goal_src/levels/village1/sage.gc [hunk 7]
```diff
@@ -5,3 +5,1422 @@
...
+      (spool-push *art-control* "sage-intro-sequence-d2" 0 self -1.0)
+      )
+    )
+   ((-> (method-of-type process-taskable play-anim) trans))
+   (none)
+   )
+  )
+
+;; definition for method 39 of type sage
+(defmethod target-close? sage ((obj sage))
```
Comment (water111) [inline goal_src/levels/village1/sage.gc:1372]: Looks like this method might actually be something like `should-display?` or `is-visible?` or something like that.

File: goal_src/levels/village1/sage.gc [hunk 8]
```diff
@@ -5,3 +5,1422 @@
...
+  (not (sages-kidnapped?))
+  )
+
+;; definition for method 41 of type sage
+;; INFO: Return type mismatch int vs none.
+(defmethod dummy-41 sage ((obj sage) (arg0 int))
```
Comment (water111) [inline goal_src/levels/village1/sage.gc:1378]: this might be called `initialize-collision`

File: goal_src/levels/village2/gambler.gc [hunk 9]
```diff
@@ -5,3 +5,522 @@
...
+                                          skel
+                                          root-channel
+                                          0
+                                          frame-group
+                                          )
+                                         )
+      num-func-seek!
+      )
+     )
+    (until (begin
```
Comment (water111) [inline goal_src/levels/village2/gambler.gc:439]: this should also get nicer with the recent changes.

File: goal_src/levels/village2/geologist.gc [hunk 10]
```diff
@@ -5,3 +5,358 @@
...
+       )
+      )
+     (-> obj draw art-group data 5)
+     )
+    )
+   )
+  )
+
+;; definition for method 31 of type geologist
+(defmethod get-3rd-art-elem geologist ((obj geologist))
```
Comment (water111) [inline goal_src/levels/village2/geologist.gc:220]: maybe just `get-art-elem`? Seems like it sometimes returns other ones.

File: goal_src/levels/village2/warrior.gc [hunk 11]
```diff
@@ -5,3 +5,371 @@
...
+     (set! (-> v1-1 settings top-plane w) (- (+ 3072.0 f0-0)))
+     )
+    0
+    )
+   )
+  (none)
+  )
+
+;; definition for method 48 of type warrior
+(defmethod TODO-RENAME-48 warrior ((obj warrior))
```
Comment (water111) [inline goal_src/levels/village2/warrior.gc:61]: maybe name this `draw-shadow`?

File: goal_src/levels/village_common/oracle.gc [hunk 12]
```diff
@@ -5,3 +5,811 @@
...
+  (set! (-> v1-1 mgeo 0) (the-as uint 1))
+  (set! (-> v1-1 lod-dist 0) 4095996000.0)
+  (set! *oracle-sg* v1-1)
+  )
+
+;; definition for method 32 of type oracle
+(defmethod play-anim! oracle ((obj oracle) (arg0 symbol))
+  (with-pp
+   (set! (-> obj talk-message) (the-as uint 260))
+   (case (current-status (-> obj tasks))
```
Comment (water111) [inline goal_src/levels/village_common/oracle.gc:44]: don't worry about it for this PR, but I think `current-status` should return a `task-status`, which should make some of these `case` statements less mysterious

File: goal_src/levels/village_common/oracle.gc [hunk 13]
```diff
@@ -5,3 +5,811 @@
...
+         :name "oracle-intro-3"
+         :index 13
+         :parts 4
+         :command-list '()
+         )
+        )
+       )
+      )
+    ((5)
+     (set! (-> obj skippable) #t)
```
Comment (water111) [inline goal_src/levels/village_common/oracle.gc:412]: it's interesting that they set this for some cutscenes, but it seems like it's not actually skippable.

File: goal_src/levels/village_common/oracle.gc [hunk 14]
```diff
@@ -5,3 +5,811 @@
...
+      (set! (-> a1-16 message) 'eval)
+      (set!
+       (-> a1-16 param 0)
+       (the-as
+        uint
+        (lambda :behavior oracle
+         ()
+         (let ((t9-0 (method-of-object (-> *part-group-id-table* 63) dummy-9)))
+          self
+          (set! (-> self part) (the-as sparticle-launch-control (t9-0)))
```
Comment (water111) [inline goal_src/levels/village_common/oracle.gc:795]: looks like this method is missing args.

File: test/decompiler/reference/engine/nav/navigate-h_REF.gc [hunk 15]
```diff
@@ -591,7 +591,7 @@
 (defmethod should-display? nav-control ((obj nav-control))
   (and
    *display-nav-marks*
-   (nonzero? (logand (-> obj flags) (nav-control-flags bit0 display-marks)))
+   (nonzero? (logand (-> obj flags) (nav-control-flags display-marks bit0)))
```
Comment (water111) [inline test/decompiler/reference/engine/nav/navigate-h_REF.gc:594]: can you double check that it still does this? I thought I fixed it so `bit0` would come first, but I might have messed it up.

**Review by water111 — COMMENTED** [review]
All the state stuff looks good! I just had a few comments on names and types.

File: test/decompiler/reference/engine/nav/navigate-h_REF.gc (hunk 15 above)
Comment (xTVaser) [reply test/decompiler/reference/engine/nav/navigate-h_REF.gc:594]: nope it seems to come second

**Pull request merged**